shared/js/content-scripts/block-cookie.js

```diff
@@ -28,72 +28,109 @@
+        document.__defineSetter__('cookie', (value) => {
+            // call the native document.cookie implementation. This will set the cookie immediately
+            // if the value is valid. We will override this set later if the policy dictates that
+            // the expiry should be changed.
+            cookieSetter.apply(document, [value])
+            try {
+                // determine the origins of the scripts in the stack
+                const scriptOrigins = (new Error().stack.split('\n')).reduce((origins, line) => {
+                    const res = line.match(lineTest)
+                    if (res && res[2]) {
+                        origins.add(new URL(res[2]).hostname)
+                    }
+                    return origins
+                }, new Set())
+                // wait for config before doing same-site tests
+                loadPolicy.then(({ tabRegisteredDomain, policy }) => {
+                    const sameSiteScript = [...scriptOrigins].every((host) => host === tabRegisteredDomain || host.endsWith(`.${tabRegisteredDomain}`))
+                    if (sameSiteScript) {
+                        // cookies set by scripts loaded on the same site as the site are not modified
+                        console.log('ignored', value, scriptOrigins)
+                        return
+                    }
+                    // extract cookie expiry from cookie string
+                    const cookieParts = value.split(';')
+                    const maxAgeIdx = cookieParts.findIndex(part => part.trim().toLowerCase().startsWith('max-age'))
+                    const expiresIdx = cookieParts.findIndex(part => part.trim().toLowerCase().startsWith('expires'))
+                    if (maxAgeIdx === -1 && expiresIdx === -1) {
+                        // session cookie
+                        return
+                    }
+                    const expiry = maxAgeIdx >= 0
+                        ? parseInt(cookieParts[maxAgeIdx].split('=')[1])
+                        : (new Date(cookieParts[expiresIdx].split('=')[1]) - new Date()) / 1000
+                    // apply cookie policy
+                    if (expiry > policy.threshold) {
+                        if (maxAgeIdx === -1) {
+                            cookieParts.push(`max-age=${policy.maxAge}`)
+                        } else {
+                            cookieParts.splice(maxAgeIdx, 1, `max-age=${policy.maxAge}`)
+                        }
+                        console.log('update cookie', cookieParts.join(';'), scriptOrigins)
+                        cookieSetter.apply(document, [cookieParts.join(';')])
+                    }
+                })
+            } catch (e) {
+                console.warn('error in cookie override', e)
             }
-
-            const cookieStr = comps.join(';')
-
-            cookieSetter.apply(document, cookieStr)
         })
+        document.__defineGetter__('cookie', cookieGetter)
```

Comment:
hm, are we doing anything with that getter? Seems like we don't need to touch it at all?

Reply:
I found if I don't do this, then `document.cookie` is `undefined`.

Reply:
TIL!

Reply:
In fact, can we leave a comment that this is the case? Someone (me in 1mo) may be tempted to remove this thinking it's useless.